### PR TITLE
make isPromise() return boolean value for null and undefined cases

### PR DESCRIPTION
**Pull request by Jayasankar-m**
- [x] By placing an `X` in the preceding checkbox, I verify that I have signed the [Contributor License Agreement](https://github.com/SeleniumHQ/selenium/blob/master/CONTRIBUTING.md#step-6-sign-the-cla)

- Currently isPromise() returns non boolean for the cases like following
isPromise(null) // returns null
isPromise(undefined) // returns undefined

- Modified to return boolean value explicitly for the cases mentioned above

---
This change is [Reviewable](https://reviewable.io/reviews/seleniumhq/selenium/6785)

### CODE REVIEW

**Review by martin770 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/seleniumhq/selenium/6785)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @Jayasankar-m)

___
*[javascript/node/selenium-webdriver/lib/promise.js, line 38 at r1](https://reviewable.io/reviews/seleniumhq/selenium/6785#-LYdhXaM6KiHlEv3CKg7:-LYdhXaM6KiHlEv3CKg8:b9hz3ej) ([raw file](https://github.com/seleniumhq/selenium/blob/d0874d3fdce4a3a4b46bfd26d626c15675482759/javascript/node/selenium-webdriver/lib/promise.js#L38)):*
> ```JavaScript
>     // Use array notation so the Closure compiler does not obfuscate away our
>     // contract.
>     return value != null
> ```

!!value to coerce the value to a boolean (instead of value != null)

Also, you should update the tests since they currently do nothing...

/test/lib/promise_test.js -- remove lines 59&60, replace with this test:

  it('isPromise', () => {
    const v = () => {};
    const x = new Promise(v, v);
    const p = createRejectedPromise('reject');
    assert.equal(true, promise.isPromise(x));
    assert.equal(true, promise.isPromise(p));
    assert.equal(false, promise.isPromise(0));
    assert.equal(false, promise.isPromise(false));
    assert.equal(false, promise.isPromise(true));
    assert.equal(false, promise.isPromise(null));
    assert.equal(false, promise.isPromise(undefined));
    assert.equal(false, promise.isPromise(''));
    assert.equal(false, promise.isPromise('promise'));
    assert.equal(false, promise.isPromise(v));
  });

**Review by martin770 — COMMENTED** [review]
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/seleniumhq/selenium/6785)* status: all files reviewed, 1 unresolved discussion (waiting on @Jayasankar-m)

___
*[javascript/node/selenium-webdriver/lib/promise.js, line 38 at r1](https://reviewable.io/reviews/seleniumhq/selenium/6785#-LYdhXaM6KiHlEv3CKg7:-LYdsBDO6s8QD_aMuYY9:bfe268f) ([raw file](https://github.com/seleniumhq/selenium/blob/d0874d3fdce4a3a4b46bfd26d626c15675482759/javascript/node/selenium-webdriver/lib/promise.js#L38)):*
<details><summary><i>Previously, martin770 (James Martin) wrote…</i></summary><blockquote>

!!value to coerce the value to a boolean (instead of value != null)

Also, you should update the tests since they currently do nothing...

/test/lib/promise_test.js -- remove lines 59&60, replace with this test:

  it('isPromise', () => {
    const v = () => {};
    const x = new Promise(v, v);
    const p = createRejectedPromise('reject');
    assert.equal(true, promise.isPromise(x));
    assert.equal(true, promise.isPromise(p));
    assert.equal(false, promise.isPromise(0));
    assert.equal(false, promise.isPromise(false));
    assert.equal(false, promise.isPromise(true));
    assert.equal(false, promise.isPromise(null));
    assert.equal(false, promise.isPromise(undefined));
    assert.equal(false, promise.isPromise(''));
    assert.equal(false, promise.isPromise('promise'));
    assert.equal(false, promise.isPromise(v));
  });
</blockquote></details>

In fact, you can completely remove the `value &&` part and just have:
```
        return (typeof value === 'object' || typeof value === 'function')
        && typeof value['then'] === 'function';
```
And it will still pass all of those tests.

**Review by Jayasankar-m — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/seleniumhq/selenium/6785)* status: all files reviewed, 1 unresolved discussion (waiting on @martin770)

___
*[javascript/node/selenium-webdriver/lib/promise.js, line 38 at r1](https://reviewable.io/reviews/seleniumhq/selenium/6785#-LYdhXaM6KiHlEv3CKg7:-LYgn_Cn1Fl1FL7_6jnI:b-h7e3il) ([raw file](https://github.com/seleniumhq/selenium/blob/d0874d3fdce4a3a4b46bfd26d626c15675482759/javascript/node/selenium-webdriver/lib/promise.js#L38)):*
<details><summary><i>Previously, martin770 (James Martin) wrote…</i></summary><blockquote>

In fact, you can completely remove the `value &&` part and just have:
```
        return (typeof value === 'object' || typeof value === 'function')
        && typeof value['then'] === 'function';
```
And it will still pass all of those tests.
</blockquote></details>

Done.
Updated as per review comments

**Review by martin770 — COMMENTED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/seleniumhq/selenium/6785)* status: :shipit: complete! all files reviewed, all discussions resolved

File: javascript/node/selenium-webdriver/lib/promise.js [hunk 1]
```diff
@@ -35,8 +35,7 @@ function isPromise(value) {
   try {
     // Use array notation so the Closure compiler does not obfuscate away our
     // contract.
-    return value
-        && (typeof value === 'object' || typeof value === 'function')
+    return (typeof value === 'object' || typeof value === 'function')
```
Comment (jleyba) [inline javascript/node/selenium-webdriver/lib/promise.js]:
```js
return value != null
    &&  (typeof value === 'object' || typeof value === 'function')
    && typeof value['then'] === 'function';
```

Comment (jleyba) [reply javascript/node/selenium-webdriver/lib/promise.js]: Note `undefined == null`

Comment (Jayasankar-m) [reply javascript/node/selenium-webdriver/lib/promise.js]: > ```js
> return value != null
>     &&  (typeof value === 'object' || typeof value === 'function')
>     && typeof value['then'] === 'function';
> ```

Updated as per review comment
Please check

**Comment by diemol** [comment]
@Jayasankar-m, would you like to continue with this PR?

**Comment by Jayasankar-m** [comment]
> @Jayasankar-m, would you like to continue with this PR?

@diemol I guess all the required changes are already there in PR. Just needs to merge.
Let me know if anything is missing

**Comment by diemol** [comment]
@Jayasankar-m, there are merge conflicts, could you please have a look?

**Comment by Jayasankar-m** [comment]
> @Jayasankar-m, there are merge conflicts, could you please have a look?

@diemol  it seems fix to the issue was made by another commit.
The tests written for this are included in the PR

**Review by diemol — APPROVED** [review]
You are right, a recent PR merge fixed this. I will merge the test, thank you.

**Pull request merged**